power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp

```diff
@@ -0,0 +1,471 @@
+        }
+        scenario_number_ = -1;
+
+        // create buffer object
+        for (MetaComponent const* const component : all_components) {
+            buffers_.push_back(count_component(batch_data, *component));
+        }
+    }
+
+    Buffer count_component(ArraySpan batch_data, MetaComponent const& component) {
+        component_key_ = component.name;
+        // count number of element of all scenarios
+        IdxVector counter(batch_size_);
+        std::vector<ArraySpan> msg_data(batch_size_);
+        for (scenario_number_ = 0; scenario_number_ != batch_size_; ++scenario_number_) {
+            msgpack::object const& scenario = batch_data[scenario_number_];
+            Idx const found_component_idx = find_key_from_map(scenario, component.name);
+            if (found_component_idx >= 0) {
+                msgpack::object const& element_array = scenario.via.map.ptr[found_component_idx].val;
+                if (element_array.type != msgpack::type::ARRAY) {
+                    throw SerializationError{"Each entry of component per scenario should be a list!\n"};
+                }
+                counter[scenario_number_] = static_cast<Idx>(element_array.via.array.size);
+                element_array >> msg_data[scenario_number_];
+            }
+        }
+        scenario_number_ = -1;
+
+        Idx const elements_per_scenario = get_elements_per_scenario(counter);
+        Idx const total_elements = // total element based on is_uniform
+            elements_per_scenario < 0 ? std::reduce(counter.cbegin(), counter.cend()) : // aggregation
+                elements_per_scenario * batch_size_;                                    // multiply
+        component_key_ = "";
+        return Buffer{.component = &component,
+                      .elements_per_scenario = elements_per_scenario,
+                      .total_elements = total_elements,
+                      .msg_data = msg_data,
+                      .data = nullptr,
+                      .indptr = {}};
+    }
+
+    bool check_uniform(IdxVector const& counter) {
+        if (batch_size_ < 2) {
+            return true;
+        }
+        return std::transform_reduce(counter.cbegin(), counter.cend() - 1, counter.cbegin() + 1, true,
+                                     std::logical_and{}, std::equal_to{});
+    }
+
+    Idx get_elements_per_scenario(IdxVector const& counter) {
+        bool const is_uniform = check_uniform(counter);
+        if (!is_uniform) {
+            return -1;
+        }
+        if (batch_size_ == 0) {
+            return 0;
+        }
+        return counter.front();
+    }
+
+    void parse_component(Buffer const& buffer) const {
+        component_key_ = buffer.component->name;
+        // handle indptr
+        if (buffer.elements_per_scenario < 0) {
+            // first always zero
+            buffer.indptr.front() = 0;
+            // accumulate sum
+            // TODO Apple Clang cannot compile transform_inclusive_scan correctly
```

Comment:
I thought we use this in the `container.hpp` as well? how do we solve it there? or is it because we need `transform_inclusive_scan` here instead of just regular `inclusive_scan`?

```suggestion
            // TODO(TonyXiang8787) Apple Clang cannot compile transform_inclusive_scan correctly
```

Reply:
In `container.hpp` we use `inclusive_scan`. That is fine with Apple Clang.